### DISCUSSION TITLE
How is one meant to update the Python venv that poetry itself uses?

### DISCUSSION THREAD

**Discussion by 13steinj** [python-poetry/poetry#8424] [Q&A] [answered]
This is a bit confusing to me. I have a number of CI pipelines and regular desktop usage that makes use of [`pyenv`](https://github.com/pyenv/pyenv) (because of transitional periods between python versions where APIs in 3.X to 3.Y do change significantly). This will be especially relevant roughly one year from now, [when "dead batteries" will be removed from the Python stdlib](https://peps.python.org/pep-0594/#id4) (which, I'd argue deserves a semantic versioning bump to 4.0; but I definitely understand why they won't do that).

So, I set up pyenv in /opt on CI pipelines to detach the dependence of a Python / PyPy / Conda / etc distribution from the operating system as much as possible. Routinely when decided, I update the python version, and if all applications can run on it, uninstall the previous one.

The problem is, poetry sets up it's own venv for it to run in. So if I install pyenv > 3.10.0 > poetry, when I upgrade 3.10 to 3.11, poetry is still on the 3.11 venv. Or even worse, I attempt to uninstall 3.10 eventually problems occur within the venv because there appears to be some level of reference to the 3.10 paths.

Loosely tangential to https://github.com/orgs/python-poetry/discussions/4693

**Comment by dimbleby** [comment]
just delete the old venv and create a new one

**Reply by 13steinj** [reply]
@dimbleby forgive me but can you elaborate? Assume I'm dumb and naive. Or maybe it's possible I was unclear.

For example, 3.11.5 has recently been released. I've done `pyenv update && pyenv install 3:latest`, and removed the `pyenv` specific one-off-venvs I have.

My poetry installation, installed with 3.11.4, created `~/.local/share/pypoetry/` and a `venv` subdirectory.

In that subdirectory, which it appears that poetry itself uses, there is a file `pyvenv.cfg, which among other things, has a direct reference to the 3.11.4 environment.

How am I to "delete the old venv and create a new one", this venv was automatically created for me upon installation of poetry.

**Reply by 13steinj** [reply]
Further, in `~/.config/pypoetry` there exists, among other files, this:

```toml
[tool.poetry]
name = "poetry-instance"
version = "1.5.1"
description = ""
authors = []
license = ""

[tool.poetry.dependencies]
python = "3.11.4"
poetry = "^1.6.1"
```

Which kinda indicates to me that poetry is not smart enough to update it's python version?

**Reply by finswimmer** [reply, marked as answer]
Poetry is installed in its own virtual environment. It is not possible to upgrade the python version of a venv. Instead the venv has to be removed and recreated with the desired interpreter.

The easiest way to do it for Poetry is to uninstall Poetry and reinstall it with the new interpreter. If you install Poetry via [pipx](https://python-poetry.org/docs/#installing-with-pipx) you can also use pipx's `reinstall` subcommand.

**Reply by 13steinj** [reply]
Okay, that works I guess.

I find it a bit odd that this isn't officially supported (by the install script)?
